lib/scholar/linear/bayesian_ridge_regression.ex

```diff
@@ -0,0 +1,320 @@
+      in a linear model. Specifically, it is the expected mean value
+      of targets for a zero-vector on input.
+      """
+    ],
+    alpha_init: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      doc: ~S"""
+      The initial value for alpha. This parameter influences the precision of the noise.
+      `:alpha` must be a non-negative float i.e. in [0, inf).
+      Defaults to 1/Var(y).      
+      """
+    ],
+    lambda_init: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0,
+      doc: ~S"""
+      The initial value for lambda. This parameter influences the precision of the weights.
+      `:lambda` must be a non-negative float i.e. in [0, inf).
+      Defaults to 1.
+      """
+    ],
+    alpha_1: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : shape parameter for the Gamma distribution prior      
+      over the alpha parameter.
+      """
+    ],
+    alpha_2: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : inverse scale (rate) parameter for the Gamma distribution prior
+      over the alpha parameter.
+      """
+    ],
+    lambda_1: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : shape parameter for the Gamma distribution prior
+      over the lambda parameter.            
+      """
+    ],
+    lambda_2: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : inverse scale (rate) parameter for the Gamma distribution prior      
+      over the lambda parameter.            
+      """
+    ],
+    eps: [
+      type: :float,
+      default: 1.0e-8,
+      doc:
+        "The convergence tolerance. When `Nx.sum(Nx.abs(coef - coef_new)) < :eps`, the algorithm is considered to have converged."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+  deftransform fit(x, y, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    opts =
+      [
+        sample_weights_flag: opts[:sample_weights] != nil
+      ] ++
+        opts
+
+    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, 1.0)
+    x_type = to_float_type(x)
+
+    sample_weights =
+      if Nx.is_tensor(sample_weights),
+        do: Nx.as_type(sample_weights, x_type),
+        else: Nx.tensor(sample_weights, type: x_type)
+
+    # handle vector types
+    # handle default alpha value, add eps to avoid division by 0
+    eps = Nx.Constants.smallest_positive_normal(x_type)
+    default_alpha = Nx.divide(1, Nx.add(Nx.variance(x), eps))
+    alpha = Keyword.get(opts, :alpha_init, default_alpha)
+    alpha = Nx.tensor(alpha, type: x_type)
+    opts = Keyword.put(opts, :alpha_init, alpha)
```

Comment:
Tensors should never be passed as options. You should always pass all tensors as arguments to the `defn` function. Your job in this function is to:

1. If any option should become a tensor, you remove it from the options and pass argument
2. All other options can be passed directly to the `defn`

Reply:
My understanding is that model [hyper parameters](https://github.com/JoaquinIglesiasTurina/scholar/blob/0006e71592812485cfd48808559d0f01259bc277/lib/scholar/linear/bayesian_ridge_regression.ex#L59), which are plain scalars, are [ok to be passed as opts](https://github.com/JoaquinIglesiasTurina/scholar/blob/0006e71592812485cfd48808559d0f01259bc277/lib/scholar/linear/bayesian_ridge_regression.ex#L182)  to `defn`.
Is that understanding correct? 

Please note that those hyperparameters need to be pased to the `while` loop. So, they get converted to tensors, there. Right?

Reply:
It depends. Everything that is passed as an option to a defn results in a different compilation. So, if you can convert something to a tensor and pass it as a tensor, that's ideally better. But that's not always possible. Sometimes the value is used as a shape, and different shapes always lead to different compilations.

> Please note that those hyperparameters need to be pased to the while loop. So, they get converted to tensors, there. Right?

Correct.